### ISSUE TITLE
`dbplyr` as `pool` dependency 

### ISSUE THREAD

**Issue by AkhilGNair** [rstudio/pool#61]
Is there any reason for `dbplyr` and to be a dependency for `pool`? It looks like `pool` is a more low level package than `dbplyr`.

The dependency chain is currently `pool` > `dbplyr` > `dplyr`.

It seems more natural to have `dbplyr` > `dplyr` + `pool`.

Would this simply involve moving the contents of [`R/dbplyr.r`](https://github.com/rstudio/pool/blob/master/R/dbplyr.R) straight into `dbplyr` (and importing the `Pool` class)?

I ask because a common pattern we use is to use `pool` and `DBI` to query a database, returning a table which is subsequently manipulated with `data.table`. We'd like to remove `dplyr` from the dependency list as 

 - It is unnecessary in our workflow 
 - It takes a very long time to install (even with `options(Ncpus = parrallel::detectCores())`)
 - The install time of `pool` is negligible
 - We use `packrat` and `Docker`, and so end up spending a lot of time unnecessarily installing `dplyr`

**Comment by AkhilNairAmey** [comment]
I've made some guesses about how I think this could work in the PRs above. The test harnesses for both packages still pass, so hopefully it doesn't break any design philosophies you had in mind when creating these packages.

**Comment by hadley** [comment]
I think a better solution would be for pool to _suggest_ dbplyr, and then use a known trick to dynamically register the methods if dbplyr is found.

**Comment by AkhilNairAmey** [comment]
Totally also happy with that. If you could link me an example of the trick above I can do my best to implement.

Thanks!

**Comment by hadley** [comment]
You can see the basic approach in https://github.com/r-lib/bench/blob/d44a32807793623d3b19da04f82c077f5adcac3b/R/zzz.R#L17-L39 — the idea is to dynamically register the methods only when (e.g.) ggplot2 is loaded (which might happen before or after the main package is loaded)

**Comment by AkhilNairAmey** [comment]
Hi @hadley - Just had a look at this and am slightly confused.

For example, with `db_commit`:

In `utils.R` we have

```
#' @importFrom dplyr db_commit
NULL
```

which is used in `dbplyr.R`

```
#' @export
#' @rdname dplyr-db-methods
db_commit.Pool <- function(con, ...) {
  db_con <- poolCheckout(con)
  on.exit(poolReturn(db_con))
  db_commit(db_con, ...)
}
```

I understand the idea about dynamically registering the method and understand I think I understand the implementation in the code you linked to.

I'm not sure I can exactly do it in this case as I can't compile `db_commit.Pool` (and so have it available to dynamically register), without importing `dplyr::db_commit`. This would of course require `dplyr` to be available.

My confusion is because the method definitions in `r-libs/bench` do not refer to the `pkg` specified in 
```
register_s3_method <- function(pkg, generic, class, fun = NULL) 
```

Apologies if I have misunderstood something here. Any advice appreciated!

**Comment by hadley** [comment]
You would no longer  import `db_commit` and no longer `@export` `db_commit.Pool`

**Comment by AkhilNairAmey** [comment]
Thanks @hadley. I've made the changes in #74. 

I added a `skip_if_not_installed` for the tests (following the example in r-lib/bench), but I'm not sure if that is sufficient. I don't know how the testing works in this case, as the dplyr tests fail on my local machine with dplyr installed.

**Comment by AkhilNairAmey** [comment]
Currently struggling to understand this behaviour:

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(pool)

pool <- dbPool(RSQLite::SQLite(), dbname = ":memory:")

# describe the type of the pool/its connections
db_desc(pool)
#> Error in UseMethod("db_desc"): no applicable method for 'db_desc' applied to an object of class "c('SQLiteConnection', 'DBIConnection', 'DBIObject')"

# use dplyr syntax to copy a table into the database
copy_to(pool, mtcars, "mtcars", temporary = FALSE)

# describe the type of the pool/its connections
db_desc(pool)
#> [1] "sqlite 3.11.1 [:memory:]"
```

**Comment by AkhilNairAmey** [comment]
@hadley All done and travis checks pass! Let me know if any other changes need to be made.. For example, I ended up adding `dplyr::check_dbplyr()` to all 18 methods which definitely did not feel completely right. I imagine the check can instead just be done in one clever place somewhere.

**Comment by AkhilGNair** [comment]
Closed in #106
